src/apply-networking-linux.py

```diff
@@ -215,32 +216,53 @@ def set_network_config_file(self, network_data, reset_to_dhcp=False):
-                    "Link could not be found " + link["ethernet_mac_address"])
-            link["os_link_name"] = os_link_name
-            links[link["id"]] = link
-
-            LOG("Apply config for link " + os_link_name)
-            base_cmd = ["ip", "link", "set", "dev", os_link_name]
-
-            mtu_cmd = base_cmd + ["mtu", link["mtu"]]
-            out, err, exit_code = execute_process(mtu_cmd, shell=False)
+    def _get_device_for_link(self, network_data, link):
+        for n_link in network_data["links"]:
+            LOG(">> %r --> %r"  %(n_link["id"], link))
+            if n_link["id"] == link:
+                return get_os_net_interface_by_mac(n_link["ethernet_mac_address"])
+        raise Exception("could not find device for link %s" % link)
+
+    def _set_link_mtu(self, link, mtu):
+        LOG("Setting MTU for link %s to %r" % (link, mtu))
+        ip_cmd = ["ip", "link", "set", "dev", link, "mtu", mtu]
+        _, err, exit_code = execute_process(ip_cmd, shell=False)
+        if exit_code:
+            raise Exception("MTU could not be set: %s" % err)
+
+    def _set_link_online(self, link):
+        ip_cmd = ["ip", "link", "set", "dev", link, "up"]
+        _, err, exit_code = execute_process(ip_cmd, shell=False)
+        if exit_code:
+            raise Exception("MTU could not be set: %s" % err)
+
+    def _flush_nic(self, link):
+        for i in ("4", "6"):
+            flush_addr_cmd = ["ip", "-%s" % i, "addr", "flush", "dev", link]
+            LOG("Executing: %s" % " ".join(flush_addr_cmd))
+            out, err, exit_code = execute_process(flush_addr_cmd, shell=False)
             if exit_code:
-                raise Exception("MTU could not be set")
+                raise Exception("IPs could not be flushed")
 
-            up_cmd = base_cmd + ["up"]
-            out, err, exit_code = execute_process(up_cmd, shell=False)
+            flush_route_cmd = ["ip", "-%s" % i, "route", "flush", "dev",
+                                          link, "scope", "global"]
+            LOG("Executing: %s" % " ".join(flush_route_cmd))
+            out, err, exit_code = execute_process(flush_route_cmd, shell=False)
             if exit_code:
-                raise Exception("Link could not be set to up state")
+                raise Exception("Routes could not be flushed")
 
+    def apply_network_config(self, network_data, reset_to_dhcp=False):
+        links = {}
+        for link in network_data["links"]:
+            link_dev = self._get_device_for_link(network_data, link["id"])
+            self._set_link_online(link_dev)
+            self._set_link_mtu(link_dev, link["mtu"])
+
+        flushed_links = set()
+        route_destinations = set()
         for network in network_data["networks"]:
-            os_link_name = links[network["link"]]["os_link_name"]
+            #os_link_name = links[network["link"]]["os_link_name"]
+            os_link_name = self._get_device_for_link(network_data, network["link"])
```

Comment:
you use this native query again because you are not sure that the link rename succeeded? doing this over and over will make the script slower. should not be a problem with one link and two networks though.